xds/internal/xdsclient/tests/ads_stream_backoff_test.go

```diff
@@ -0,0 +1,438 @@
+	streamCloseCh := make(chan struct{}, 1)  // ADS stream is closed.
+	ldsResourcesCh := make(chan []string, 1) // Listener resource names in the discovery request.
+	backoffCh := make(chan struct{}, 1)      // Backoff after stream failure.
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Create an xDS management server that returns RPC errors.
+	streamErr := errors.New("ADS stream error")
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{
+		OnStreamRequest: func(_ int64, req *v3discoverypb.DiscoveryRequest) error {
+			// Push the requested resource names on to a channel.
+			if req.GetTypeUrl() == version.V3ListenerURL {
+				t.Logf("Received LDS request for resources: %v", req.GetResourceNames())
+				select {
+				case ldsResourcesCh <- req.GetResourceNames():
+				default:
+				}
+			}
+			// Return an error everytime a request is sent on the stream. This
+			// should cause the transport to backoff before attempting to
+			// recreate the stream.
+			return streamErr
+		},
+		// Push on a channel whenever the stream is closed.
+		OnStreamClosed: func(int64, *v3corepb.Node) {
+			select {
+			case streamCloseCh <- struct{}{}:
+			default:
+			}
+		},
+	})
+
+	// Override the backoff implementation to push on a channel that is read by
+	// the test goroutine.
+	streamBackoff := func(v int) time.Duration {
+		select {
+		case backoffCh <- struct{}{}:
+		default:
+		}
+		return 0
+	}
+
+	// Create an xDS client with bootstrap pointing to the above server.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+	client := createXDSClientWithBackoff(t, bc, streamBackoff)
+
+	// Register a watch for a listener resource.
+	const listenerName = "listener"
+	lw := newListenerWatcher()
+	ldsCancel := xdsresource.WatchListener(client, listenerName, lw)
+	defer ldsCancel()
+
+	// Verify that an ADS stream is created and an LDS request with the above
+	// resource name is sent.
+	if err := waitForResourceNames(ctx, t, ldsResourcesCh, []string{listenerName}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that the received stream error is reported to the watcher.
+	u, err := lw.updateCh.Receive(ctx)
+	if err != nil {
+		t.Fatal("Timeout when waiting for an error callback on the listener watcher")
+	}
+	gotErr := u.(listenerUpdateErrTuple).err
+	if !strings.Contains(gotErr.Error(), streamErr.Error()) {
+		t.Fatalf("Received stream error: %v, wantErr: %v", gotErr, streamErr)
+	}
+
+	// Verify that the stream is closed.
+	select {
+	case <-streamCloseCh:
+	case <-ctx.Done():
+		t.Fatalf("Timeout waiting for stream to be closed after an error")
+	}
+
+	// Verify that the ADS stream backs off before recreating the stream.
+	select {
+	case <-backoffCh:
+	case <-ctx.Done():
+		t.Fatalf("Timeout waiting for ADS stream to backoff after stream failure")
+	}
```

Comment:
Is this flaky and select on line 139 flaky with respect to extra sends on the send path due to the default? I guess it's flaky in the sense that it doesn't verify steam closure/backoff exactly once. Maybe that's ok since backoff is 0 so it's non deterministic anyway. Can this assertion be exact? I don't feel strongly about this, just musing.

Reply:
For the purposes of this test, I only care that the stream backoff happens once after stream failure. The code that handles backoff is the same irrespective of whether it is the first stream failure or the Nth stream failure (except of course when a message is received on the stream, the backoff state is reset). So, I feel it makes total sense to do non-blocking channel writes here. 

If you feel uncomfortable about this, I *could* switch this to use a `grpcsync.Event` where you can fire the event multiple times, but all we care is if the event fired once at all. Let me know. Thanks.

Reply:
I don't feel strongly about grpcsync.Event vs. this, those seem pretty equivalent. My point is I feel like we can make these assertions stronger and expect the exact amount of times backoff will be called rather than being ok with any number of times 1 to infinity, especially if we incorporate some channels that block until the test goroutine signals on them. If you can make the backoff deterministic, why not switch the assertion to exact number of times backoff is called? I feel like exact number is a stronger assertion than called at least once.

Reply:
I switched to have a case in the `select` for `ctx.Done()`. This I feel is required because during shutdown, connections go down, and there could be backoff happening in the xDS client based on the ordering of close events.

Reply:
Ok that sgtm. Now it asserts 1 exact, and does the right thing on races in shutdown. Thanks.